File: src/com/l2kt/gameserver/network/clientpackets/SetPrivateStoreListBuy.java
package com.l2kt.gameserver.network.clientpackets;

import com.l2kt.Config;
import com.l2kt.gameserver.model.actor.instance.Player;
import com.l2kt.gameserver.model.tradelist.TradeList;
import com.l2kt.gameserver.model.zone.ZoneId;
import com.l2kt.gameserver.network.SystemMessageId;
import com.l2kt.gameserver.network.serverpackets.PrivateStoreManageListBuy;
import com.l2kt.gameserver.network.serverpackets.PrivateStoreMsgBuy;
import com.l2kt.gameserver.taskmanager.AttackStanceTaskManager;

public final class SetPrivateStoreListBuy extends L2GameClientPacket
{
	private static final int BATCH_LENGTH = 16; // length of one item
	
	private Item[] _items = null;
	
	@Override
	protected void readImpl()
	{
		int count = readD();
		if (count < 1 || count > Config.MAX_ITEM_IN_PACKET || count * BATCH_LENGTH != _buf.remaining())
			return;
		
		_items = new Item[count];
		for (int i = 0; i < count; i++)
		{
			int itemId = readD();
			readH(); // TODO analyse this
			readH(); // TODO analyse this
			int cnt = readD();
			int price = readD();
			
			if (itemId < 1 || cnt < 1 || price < 0)
			{
				_items = null;
				return;
			}
			_items[i] = new Item(itemId, cnt, price);
		}
	}
	
	@Override
	protected void runImpl()
	{
		Player player = getClient().getActiveChar();
		if (player == null)
			return;
		
		if (_items == null)
		{
			player.setStoreType(Player.StoreType.NONE);
			player.broadcastUserInfo();
			player.sendPacket(new PrivateStoreManageListBuy(player));
			return;
		}
		
		if (!player.getAccessLevel().allowTransaction())
		{
			player.sendPacket(SystemMessageId.YOU_ARE_NOT_AUTHORIZED_TO_DO_THAT);
			return;
		}
		
		if (AttackStanceTaskManager.getInstance().isInAttackStance(player) || (player.isCastingNow() || player.isCastingSimultaneouslyNow()) || player.isInDuel())
		{
			player.sendPacket(SystemMessageId.YOU_ARE_NOT_AUTHORIZED_TO_DO_THAT);
			player.sendPacket(new PrivateStoreManageListBuy(player));
			return;
		}
		
		if (player.isInsideZone(ZoneId.NO_STORE))
		{
			player.sendPacket(SystemMessageId.NO_PRIVATE_STORE_HERE);
			player.sendPacket(new PrivateStoreManageListBuy(player));
			return;
		}
		
		TradeList tradeList = player.getBuyList();
		tradeList.clear();
		
		// Check maximum number of allowed slots for pvt shops
		if (_items.length > player.getPrivateBuyStoreLimit())
		{
			player.sendPacket(SystemMessageId.YOU_HAVE_EXCEEDED_QUANTITY_THAT_CAN_BE_INPUTTED);
			player.sendPacket(new PrivateStoreManageListBuy(player));
			return;
		}
		
		int totalCost = 0;
		for (Item i : _items)
		{
			if (!i.addToTradeList(tradeList))
			{
				player.sendPacket(SystemMessageId.EXCEEDED_THE_MAXIMUM);
				player.sendPacket(new PrivateStoreManageListBuy(player));
				return;
			}
			
			totalCost += i.getCost();
			if (totalCost > Integer.MAX_VALUE)
			{
				player.sendPacket(SystemMessageId.EXCEEDED_THE_MAXIMUM);
				player.sendPacket(new PrivateStoreManageListBuy(player));
				return;
			}
		}
		
		// Check for available funds
		if (totalCost > player.getAdena())
		{
			player.sendPacket(SystemMessageId.THE_PURCHASE_PRICE_IS_HIGHER_THAN_MONEY);
			player.sendPacket(new PrivateStoreManageListBuy(player));
			return;
		}
		
		player.sitDown();
		player.setStoreType(Player.StoreType.BUY);
		player.broadcastUserInfo();
		player.broadcastPacket(new PrivateStoreMsgBuy(player));
	}
	
	private static class Item
	{
		private final int _itemId, _count, _price;
		
		public Item(int id, int num, int pri)
		{
			_itemId = id;
			_count = num;
			_price = pri;
		}
		
		public boolean addToTradeList(TradeList list)
		{
			if ((Integer.MAX_VALUE / _count) < _price)
				return false;
			
			list.addItemByItemId(_itemId, _count, _price);
			return true;
		}
		
		public long getCost()
		{
			return _count * _price;
		}
	}
}

File: src/com/l2kt/gameserver/network/clientpackets/SetPrivateStoreListBuy.kt
package com.l2kt.gameserver.network.clientpackets

import com.l2kt.Config
import com.l2kt.gameserver.model.actor.instance.Player
import com.l2kt.gameserver.model.tradelist.TradeList
import com.l2kt.gameserver.model.zone.ZoneId
import com.l2kt.gameserver.network.SystemMessageId
import com.l2kt.gameserver.network.serverpackets.PrivateStoreManageListBuy
import com.l2kt.gameserver.network.serverpackets.PrivateStoreMsgBuy
import com.l2kt.gameserver.taskmanager.AttackStanceTaskManager

class SetPrivateStoreListBuy : L2GameClientPacket() {

    private var _items: MutableList<Item> = mutableListOf()

    override fun readImpl() {
        val count = readD()
        if (count < 1 || count > Config.MAX_ITEM_IN_PACKET || count * BATCH_LENGTH != _buf.remaining())
            return

        for (i in 0 until count) {
            val itemId = readD()
            readH() // TODO analyse this
            readH() // TODO analyse this
            val cnt = readD()
            val price = readD()

            if (itemId < 1 || cnt < 1 || price < 0) {
                _items = mutableListOf()
                return
            }
            _items.add(i, Item(itemId, cnt, price))
        }
    }

    override fun runImpl() {
        val player = client.activeChar ?: return

        if (_items.isEmpty()) {
            player.storeType = Player.StoreType.NONE
            player.broadcastUserInfo()
            player.sendPacket(PrivateStoreManageListBuy(player))
            return
        }

        if (!player.accessLevel.allowTransaction()) {
            player.sendPacket(SystemMessageId.YOU_ARE_NOT_AUTHORIZED_TO_DO_THAT)
            return
        }

        if (AttackStanceTaskManager.isInAttackStance(player) || player.isCastingNow || player.isCastingSimultaneouslyNow || player.isInDuel) {
            player.sendPacket(SystemMessageId.YOU_ARE_NOT_AUTHORIZED_TO_DO_THAT)
            player.sendPacket(PrivateStoreManageListBuy(player))
            return
        }

        if (player.isInsideZone(ZoneId.NO_STORE)) {
            player.sendPacket(SystemMessageId.NO_PRIVATE_STORE_HERE)
            player.sendPacket(PrivateStoreManageListBuy(player))
            return
        }

        val tradeList = player.buyList
        tradeList.clear()

        // Check maximum number of allowed slots for pvt shops
        if (_items.size > player.privateBuyStoreLimit) {
            player.sendPacket(SystemMessageId.YOU_HAVE_EXCEEDED_QUANTITY_THAT_CAN_BE_INPUTTED)
            player.sendPacket(PrivateStoreManageListBuy(player))
            return
        }

        var totalCost = 0
        for (i in _items) {
            if (!i.addToTradeList(tradeList)) {
                player.sendPacket(SystemMessageId.EXCEEDED_THE_MAXIMUM)
                player.sendPacket(PrivateStoreManageListBuy(player))
                return
            }

            totalCost += i.cost.toInt()
            if (totalCost > Integer.MAX_VALUE) {
                player.sendPacket(SystemMessageId.EXCEEDED_THE_MAXIMUM)
                player.sendPacket(PrivateStoreManageListBuy(player))
                return
            }
        }

        // Check for available funds
        if (totalCost > player.adena) {
            player.sendPacket(SystemMessageId.THE_PURCHASE_PRICE_IS_HIGHER_THAN_MONEY)
            player.sendPacket(PrivateStoreManageListBuy(player))
            return
        }

        player.sitDown()
        player.storeType = Player.StoreType.BUY
        player.broadcastUserInfo()
        player.broadcastPacket(PrivateStoreMsgBuy(player))
    }

    private class Item(private val _itemId: Int, private val _count: Int, private val _price: Int) {

        val cost: Long
            get() = (_count * _price).toLong()

        fun addToTradeList(list: TradeList): Boolean {
            if (Integer.MAX_VALUE / _count < _price)
                return false

            list.addItemByItemId(_itemId, _count, _price)
            return true
        }
    }

    companion object {
        private const val BATCH_LENGTH = 16 // length of one item
    }
}